sigmastate/src/main/scala/sigmastate/interpreter/Interpreter.scala

```diff
@@ -156,70 +201,186 @@ trait Interpreter extends ScorexLogging {
   def reduceToCrypto(context: CTX, exp: Value[SType]): Try[ReductionResult] =
     reduceToCrypto(context, Interpreter.emptyEnv, exp)
 
+  /** This method uses the new JIT costing with direct ErgoTree execution. It is used in
+    * both prover and verifier to compute SigmaProp value.
+    * As the first step the cost of computing the `exp` expression in the given context is
+    * estimated.
+    * If cost is above limit then exception is returned and `exp` is not executed
+    * else `exp` is computed in the given context and the resulting SigmaBoolean returned.
+    *
+    * @param context        the context in which `exp` should be executed
+    * @param env            environment of system variables used by the interpreter internally
+    * @param exp            expression to be executed in the given `context`
+    * @return result of script reduction
+    * @see `ReductionResult`
+    */
+  def reduceToCryptoJITC(context: CTX, env: ScriptEnv, exp: SigmaPropValue): Try[ReductionResult] = Try {
+    implicit val vs = context.validationSettings
+    trySoftForkable[ReductionResult](whenSoftFork = WhenSoftForkReductionResult(context.initCost)) {
+
+      val (resProp, cost) = {
+        val ctx = context.asInstanceOf[ErgoLikeContext]
+          .withInitCost(context.initCost * 10)   // adjust for Evaluator cost units scale
```

Comment:
This is because JIT costs are more accurate and I found that profiler measurements are quite stable. When microsecond scale is used then the minimum cost of 1 microsecond is too rough for cheap operations. 
When such operations (like ConstantPlaceholder, Constant, AddToEnvironment, etc) are called in a loops, then the cost estimation becomes less accurate. Basically the cost model doesn't scale with data sizes. 
For this reason I started to use 100 nanoseconds as the minimal unit of cost, hence scaling with respect to the old (v4.x) microsecond scale.

Reply:
Cost unit is synthetic and not related to time units (which are hardware-dependent). 

Scaling is confusing and I think could be fixed in more elegant way.

Reply:
Which way?

Reply:
@aslesarenko if two units of account are needed, you need to split them around all the code with e.g. type aliases, so something like:

type OldCost = Long
type NewCost = Long

and then make explicit declarations (e.g. FoldOpCost = 10: OldCost) for ALL the costs 

as messing up with different units decorated in the same way is utterly confusing for a reader

Reply:
1) I'm not going to touch the old code in any way. And I don't recommend anyone to do this. So, I suggest the old cost to remain Long.
2) Introducing `type NewCost = Long` doesn't change anything typechecker-wise.
3) I will see if it will be easy to use tagged type JitCost for JITC values and whether is will incur noticable overhead.

Reply:
This is not about typechecking, rather, documentation. 
Another option is to put new costs into one table (NewCosts), old into another.
But you had some objection about externalizing costs into a table before

Reply:
Yes, it is better to keep all the metadata (like cost parameters and others) about each operation close to the implementation (i.e. eval method). Then any table can be generated from that metadata by writing a simple generator. I did this to generate all the tables of ErgoTree specification.

Reply:
If you think implementing tagged type JitCost is necessary, I would do it in a separate PR and AFTER adding test vectors for the new costing. 
This will prove, in particular, that no bugs were introduced as part of the refactoring.

Reply:
Please describe somewhere how the table with all the costs can be generated.
Also, you mean that this PR will be merged after some other one implemented? ok

Reply:
I checked costs again and it seems they can be rounded to a closest value which can be divided by 10, thus old costing precision can be used:

* most of places with FixedCost(1) are just about ops not being implemented. Better to mark such places with a special FixedCostNotImplemented class.
* For data chunks, cost must be higher to penalize for on-chain data storage for more. 
* For a cost like 5 or 15, no big deal to raise it to 10/20
* For a couple of other places, no big deal to raise from 1 to 10 also

Reply:
Current costs are turned against actual execution times and also checked many times against AOT cost to not exceed them for all existing blocks (I spent quite a time for this). 
Decreasing precision will invalidate all this work done, and I don't understand the motivation, why you want to go from better estimation to worse estimation. 
If it were a simple rollback, then ok, it can be done, even though I don't understant why. 
But this is not a rollback, the new costs should be turned and analysed again, who will do this?

What is the problem with scaling, lets maybe solve it instead?

Reply:
The main problem the thread started with is code readability (different units used with just one small comment in interpreter a reader has a little chance to find quickly). 
In case of data chunks (and maybe ops which assume that data are stored on-chain), another problem is that cost should reflect not only calculations but also disincentivize storing data on chain (not in so harsh way as done in Eth but still)

Reply:
The current cost are already disincentivize large collections.  For readability it can be solved by using JitCost tagged type over Long, with the corresponding arithmetic operations over JitCost values. So scaling can be encapsulated to JitCost -> Long and Long -> JitCost operations.

Reply:
How storing data is disincentivized if you tracked execution times only ?

For readability, https://github.com/ScorexFoundation/sigmastate-interpreter/issues/752

Reply:
The cost of loop operations is approximated by PerItemCost descriptor, which defines cost per chunk. The parameters are selected conservatively, so large collections will get bigger estimation error (i.e. more conservative).
Also collection operations like map, filter, exists, involve lambdas, which are executed in a loop. 
Since each lambda have its own conservative estimation error, this error got multiplied by the length of collection.